rust_src/src/casetab.rs

```diff
@@ -3,100 +3,138 @@
-            map_char_table(Some(shuffle), Qnil, table, up);
-            set_char_table_extras(table, 0, up);
+            let up_table = make_char_table(sym_case_table, Qnil);
+            up = LispObject::from(up_table);
+            unsafe {
+                map_char_table(Some(set_identity), Qnil, table, up);
+                map_char_table(Some(shuffle), Qnil, table, up);
+            }
+            self.set_extras(0, up);
         }
 
-        if canon.is_nil() {
-            canon = Fmake_char_table(Qcase_table, Qnil);
-            set_char_table_extras(table, 1, canon);
-            map_char_table(Some(set_canon), Qnil, table, table);
-        }
+        let mut canon_table = if canon.is_nil() {
+            let tem = make_char_table(sym_case_table, Qnil);
+            canon = LispObject::from(tem);
+            self.set_extras(1, canon);
+            unsafe {
+                map_char_table(Some(set_canon), Qnil, table, table);
+            }
+            tem
+        } else {
+            canon.into()
+        };
 
         if eqv.is_nil() {
-            eqv = Fmake_char_table(Qcase_table, Qnil);
-            map_char_table(Some(set_identity), Qnil, canon, eqv);
-            map_char_table(Some(shuffle), Qnil, canon, eqv);
-            set_char_table_extras(table, 2, eqv);
+            let eqv_table = make_char_table(sym_case_table, Qnil);
+            eqv = eqv_table.into();
+            unsafe {
+                map_char_table(Some(set_identity), Qnil, canon, eqv);
+                map_char_table(Some(shuffle), Qnil, canon, eqv);
+            }
+            self.set_extras(2, eqv);
         }
 
         // This is so set_image_of_range_1 in regex.c can find the EQV table.
-        set_char_table_extras(canon, 2, eqv);
+        canon_table.set_extras(2, eqv);
     }
+}
 
-    if standard {
-        unsafe {
-            Vascii_downcase_table = table;
-            Vascii_upcase_table = up;
-            Vascii_canon_table = canon;
-            Vascii_eqv_table = eqv;
-        }
-    } else {
-        let mut buffer: LispBufferRef = current_buffer().into();
-        buffer.downcase_table_ = table;
-        buffer.upcase_table_ = up;
-        buffer.case_canon_table_ = canon;
-        buffer.case_eqv_table_ = eqv;
+impl LispObject {
+    pub fn as_case_table(self) -> Option<LispCaseTable> {
+        self.into()
+    }
+}
+
+impl From<LispObject> for LispCaseTable {
+    fn from(obj: LispObject) -> Self {
+        obj.as_case_table()
+            .unwrap_or_else(|| wrong_type!(Qcase_table_p, obj))
     }
+}
 
-    table
+// Allow for the transformation of char tables into case tables when their
+// `purpose` field is set to `Qcase_table`. Do not check the `extras` fields.
+// This way a new char table can be transformed into a case table.
+impl From<LispObject> for Option<LispCaseTable> {
+    fn from(obj: LispObject) -> Self {
+        Option::<LispCharTableRef>::from(obj).and_then(LispCaseTable::from_char_table)
```

Comment:
It feels like it would make sense to require `is_proper` to be true for the conversion to succeed, I think the "unchecked" version should have a separate method.